rcl_yaml_param_parser/src/parse.c

```diff
@@ -0,0 +1,832 @@
+          if (NULL == node_name_ns) {
+            ret = RCUTILS_RET_BAD_ALLOC;
+            break;
+          }
+
+          ret = find_node(node_name_ns, params_st, node_idx);
+          allocator.deallocate(node_name_ns, allocator.state);
+          if (RCUTILS_RET_OK != ret) {
+            break;
+          }
+
+          ret = rem_name_from_ns(ns_tracker, NS_TYPE_NODE, allocator);
+          if (RCUTILS_RET_OK != ret) {
+            RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+              "Internal error adding node namespace at line %d", line_num);
+            break;
+          }
+          /// Bump the map level to PARAMS
+          (*map_level)++;
+        }
+      }
+      break;
+    case MAP_PARAMS_LVL:
+      {
+        char * parameter_ns = NULL;
+        char * param_name = NULL;
+
+        /// If it is a new map, the previous key is param namespace
+        if (true == *is_new_map) {
+          parameter_ns = params_st->params[*node_idx].parameter_names[*parameter_idx];
+          if (NULL == parameter_ns) {
+            RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+              "Internal error creating param namespace at line %d", line_num);
+            ret = RCUTILS_RET_ERROR;
+            break;
+          }
+          ret = replace_ns(
+            ns_tracker, parameter_ns, (ns_tracker->num_parameter_ns + 1U),
+            NS_TYPE_PARAM, allocator);
+          if (RCUTILS_RET_OK != ret) {
+            RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+              "Internal error replacing namespace at line %d", line_num);
+            ret = RCUTILS_RET_ERROR;
+            break;
+          }
+          *is_new_map = false;
+        }
+
+        // Guard against adding more than the maximum allowed parameters
+        if (params_st->params[*node_idx].num_params >= MAX_NUM_PARAMS_PER_NODE) {
+          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+            "Exceeded maximum allowed number of parameters for a node (%d)",
+            MAX_NUM_PARAMS_PER_NODE);
+          ret = RCUTILS_RET_ERROR;
+          break;
+        }
+
+        /// Add a parameter name into the node parameters
+        parameter_ns = ns_tracker->parameter_ns;
+        if (NULL == parameter_ns) {
+          ret = find_parameter(*node_idx, value, params_st, parameter_idx);
+          if (ret != RCUTILS_RET_OK) {
+            break;
+          }
+        } else {
+          ret = find_parameter(*node_idx, parameter_ns, params_st, parameter_idx);
+          if (ret != RCUTILS_RET_OK) {
+            break;
+          }
+
+          const size_t params_ns_len = strlen(parameter_ns);
+          const size_t param_name_len = strlen(value);
+          const size_t tot_len = (params_ns_len + param_name_len + 2U);
+
+          param_name = allocator.zero_allocate(1U, tot_len, allocator.state);
+          if (NULL == param_name) {
+            ret = RCUTILS_RET_BAD_ALLOC;
+            break;
+          }
+
+          memmove(param_name, parameter_ns, params_ns_len);
```

Comment:
@brawner perhaps unrelated, we can use `memcpy` instead.

Reply:
Switched to memcpy